core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java

```diff
@@ -390,4 +414,98 @@ private static boolean isSplitLocal(List<HostAddress> splitAddresses, HostAddres
         }
         return false;
     }
+
+    static class QueueSizeAdjuster
```

Comment:
The field `nodeMap` is type of `AtomicReference<Supplier<NodeMap>>` and it is updated from `lockdownNodes` that is called from `SourcePartitionedSchedule.schedule`. From that method the `computeAssigments` (and your operations on `QueueSizeAdjuster`)  is called as well. 
Using `AtomicReference` could point that that method can be run in parallel. Maybe `QueueSizeAdjuster` should  be thread safe as well?

Reply:
`SourcePartitionedSchedule::schedule` is synchronized. For my understanding, `UniformNodeSelector::computeAssignments` should not be run in parallel, as we don't share any of assignments statistics during of assignment loop (for example, see `NodeAssignmentStats::addAssignedSplit`). `AtomicReference` is for capturing a snapshot of a variable that could be altered by other thread. Adding possibility to make `QueueSizeAdjuster` looks like unnecessary work.